source/kernel/trace_inode.c

```diff
@@ -0,0 +1,231 @@
+#include "trace_inode.h"
+
+#include <linux/atomic.h>
+#include <linux/dcache.h>
+#include <linux/errno.h>
+#include <linux/fs.h>
+#include <linux/hashtable.h>
+#include <linux/list.h>
+#include <linux/slab.h>
+#include "io_trace.h"
+#include "iotrace_event.h"
+#include "trace_env_kernel.h"
+
+#ifndef DEBUG
+#define DEBUG 0
+#endif
+
+#if 1 == DEBUG
+#define debug(format, ...)                                               \
+    printk(KERN_INFO "[iotrace][inode cache] %s " format "\n", __func__, \
+           ##__VA_ARGS__)
+#else
+#define debug(format, ...)
+#endif
+
+#define CACHE_ENTRIES 4096ULL
+#define HASH_ENTRIES CACHE_ENTRIES / 4ULL
+
+struct cache_entry {
+    struct list_head lru;
+    struct hlist_node hash;
+    uint64_t inode_id;
+    dev_t device_id;
+};
+
+struct iotrace_inode {
+    int cpu;
+    DECLARE_HASHTABLE(hash_table, ilog2(HASH_ENTRIES));
+    struct list_head list_free;
+    struct list_head list_lru;
+    struct cache_entry entries[CACHE_ENTRIES];
+};
+
+int iotrace_inode_create(iotrace_inode_t *_iotrace_inode, int cpu) {
+    int i;
+    struct iotrace_inode *iotrace_inode;
+
+    debug();
+
+    *_iotrace_inode = NULL;
+    iotrace_inode = vzalloc_node(sizeof(*iotrace_inode), cpu_to_node(cpu));
+    if (!iotrace_inode) {
+        return -ENOMEM;
+    }
+
+    iotrace_inode->cpu = cpu;
+    hash_init(iotrace_inode->hash_table);
+    INIT_LIST_HEAD(&iotrace_inode->list_free);
+    INIT_LIST_HEAD(&iotrace_inode->list_lru);
+
+    /* Initialize free list */
+    for (i = 0; i < ARRAY_SIZE(iotrace_inode->entries); i++) {
+        struct cache_entry *entry = &iotrace_inode->entries[i];
+        list_add(&entry->lru, &iotrace_inode->list_free);
+    }
+
+    *_iotrace_inode = iotrace_inode;
+    return 0;
+}
+
+void iotrace_inode_destroy(iotrace_inode_t *iotrace_inode) {
+    debug();
+
+    if (*iotrace_inode) {
+        vfree(*iotrace_inode);
+        *iotrace_inode = NULL;
+    }
+}
+
+static uint64_t _hash_index(struct inode *inode) {
+    return inode->i_ino % HASH_ENTRIES;
```

Comment:
Isn't this hashing taken care of by hashtable implementation?  I think you can use values out of scope as indexes.
https://kernelnewbies.org/FAQ/Hashtables